crates/oxc_query/src/properties.rs

```diff
@@ -0,0 +1,584 @@
+        }
+    }
+}
+
+pub(super) fn resolve_import_ast_property<'a, 'b: 'a>(
+    contexts: ContextIterator<'a, Vertex<'b>>,
+    property_name: &str,
+    resolve_info: &ResolveInfo,
+) -> ContextOutcomeIterator<'a, Vertex<'b>, FieldValue> {
+    match property_name {
+        "from_path" => resolve_import_property(contexts, property_name, resolve_info),
+        _ => {
+            unreachable!(
+                "attempted to read unexpected property '{property_name}' on type 'ImportAST'"
+            )
+        }
+    }
+}
+
+pub(super) fn resolve_interface_extend_property<'a, 'b: 'a>(
+    contexts: ContextIterator<'a, Vertex<'b>>,
+    property_name: &str,
+    _resolve_info: &ResolveInfo,
+) -> ContextOutcomeIterator<'a, Vertex<'b>, FieldValue> {
+    match property_name {
+        "str" => interface_extend_implem(contexts),
+        _ => {
+            unreachable!(
+                "attempted to read unexpected property '{property_name}' on type 'InterfaceExtend'"
+            )
+        }
+    }
+}
+
+pub(super) fn resolve_jsxattribute_property<'a, 'b: 'a>(
+    contexts: ContextIterator<'a, Vertex<'b>>,
+    property_name: &str,
+    _resolve_info: &ResolveInfo,
+) -> ContextOutcomeIterator<'a, Vertex<'b>, FieldValue> {
+    match property_name {
+        "name" => resolve_property_with(contexts, |v| {
+            jsx_attribute_name_to_string(
+                &v.as_jsx_attribute()
+                    .unwrap_or_else(|| {
+                        panic!("expected to have a jsxattribute vertex, instead have: {v:#?}")
+                    })
+                    .name,
+            )
+            .into()
+        }),
+        "value_as_constant_string" => resolve_property_with(contexts, |v| {
+            let attr = v.as_jsx_attribute().unwrap_or_else(|| {
+                panic!("expected to have a jsxattribute vertex, instead have: {v:#?}")
+            });
+            jsx_attribute_to_constant_string(attr).map_or_else(|| FieldValue::Null, Into::into)
+        }),
+        _ => {
+            unreachable!(
+                "attempted to read unexpected property '{property_name}' on type 'JSXAttribute'"
+            )
+        }
+    }
+}
+
+pub(super) fn resolve_jsxelement_property<'a, 'b: 'a>(
+    contexts: ContextIterator<'a, Vertex<'b>>,
+    property_name: &str,
+    _resolve_info: &ResolveInfo,
+) -> ContextOutcomeIterator<'a, Vertex<'b>, FieldValue> {
+    match property_name {
+        "as_constant_string" => resolve_property_with(contexts, |v| {
+            v.as_constant_string().map_or(FieldValue::Null, Into::into)
+        }),
+        "child_count" => resolve_property_with(contexts, |v| {
+            (v.as_jsx_element()
+                .unwrap_or_else(|| {
+                    panic!("expected to have a jsxelement vertex, instead have: {v:#?}")
+                })
+                .element
+                .children
+                .len() as u64)
+                .into()
+        }),
```

Comment:
Another opportunity to use one of those helper macros I mentioned:
```suggestion
        "child_count" => resolve_property_with(contexts, field_property!(as_jsx_element, element, {
            (element.children.len() as u64).into()
        }),
```

In this case the macro grabs the `.element` field and gives it to you in the block to further manipulate.

When I write adapters, I tend to find this style preferable because it's less repetitive (`as_vertex().expect(...)` etc.), but you do you.

Reply:
done in [4f4b165](https://github.com/web-infra-dev/oxc/pull/599/commits/4f4b165ec5add96b3daf5a4a05f25680c4fe186a)